### PR TITLE
feat: add transactionAsyncLocalStorage option to opt in to automatically setting session on all transactions

### PR DESCRIPTION
**Pull request by vkarpov15**
Backport #14583 to 7.x
Re: #13889

**Summary**

Add `transactionAsyncLocalStorage` option to 7.x. Should work fine because Node v14.x has "experimental" support for the same AsyncLocalStorage API that became stable in 16.x.

**Examples**

### CODE REVIEW

File: docs/transactions.md
Comment (hasezoey) [inline docs/transactions.md:91]:
```suggestion
Mongoose 7.8 is able to set the `session` operation on all operations within a `Connection.prototype.transaction()` executor function using Node's [AsyncLocalStorage API](https://nodejs.org/api/async_context.html#class-asynclocalstorage).
```

File: types/mongooseoptions.d.ts
Comment (hasezoey) [inline types/mongooseoptions.d.ts:207]:
```suggestion
     * Set to true to make Mongoose use Node.js' built-in AsyncLocalStorage (Added in: v13.10.0, v12.17.0; Stable since 16.4.0)
```

**Review by hasezoey — CHANGES_REQUESTED** [review]
Code looks good to me, though i would recommend to add a note to the documentation that it makes use of a experimental nodejs API below `16.4.0` (as added in the types by me and listed [under `AsyncLocalStorage`'s History](https://nodejs.org/api/async_context.html#class-asynclocalstorage)).

also i just noticed that option `transactionAsyncLocalStorage` (in latest docs) only returns a result for `docs/transactions.html`, maybe it should be mentioned as a possible option for `mongoose.set`.

aside from that, there are some minor doc suggestions that should be done.

**Review by hasezoey — COMMENTED** [review]
LGTM, would wait until #14744 is merged (and included here) before approving

**Comment by vkarpov15** [comment]
@hasezoey I backported #14744 in e909e0b, can you take a look?

**Review by hasezoey — APPROVED** [review]

**Pull request merged**
